sunscreen_tfhe/src/ops/encryption/rlwe_encryption.rs

```diff
@@ -0,0 +1,198 @@
+use crate::{
+    dst::FromMutSlice,
+    entities::{GlweCiphertext, GlweSecretKeyRef, Polynomial, PolynomialRef, RlwePublicKeyRef},
+    ops::encryption::encrypt_glwe_ciphertext_secret,
+    polynomial::{polynomial_add_assign, polynomial_external_mad},
+    rand::{binary_torus_polynomial, normal_torus_polynomial},
+    scratch::allocate_scratch_ref,
+    GlweDef, PlaintextBits, Torus, TorusOps,
+};
+
+/// The randomness used to generate a public-key RLWE encryption of a message.
+/// Useful when generating zero-knowledge proofs.
+///
+/// # Security
+/// This information must remain private. Sharing this information compromises the
+/// security of the encrypted message.
+///
+/// # Remarks
+/// See [`rlwe_encrypt_public`] for an explanation of the RLWE public-key encryption
+/// algorithm and the significance of `e_1`, `e_2`, and `u`.
+pub struct RlwePublicEncryptionRandomness<S>
+where
+    S: TorusOps,
+{
+    /// The normal e_1 term.
+    pub e0: Polynomial<Torus<S>>,
+
+    /// The normal e_2 term.
+    pub e1: Polynomial<Torus<S>>,
+
+    /// The binary u term.
+    pub u: Polynomial<S>,
+}
+
+/// Generate an RLWE public key into `public_key` for the given `secret_key`.
+///
+/// # Remarks
+/// RLWE is a special case of GLWE where N is a power of 2 as normal, but size is 1.
+///
+/// An RLWE public key is simply an secret-key RLWE encryption of the zero polynomial.
+///
+/// # Panics
+/// If `glwe.dim.size != 1`.
+pub fn rlwe_generate_public_key<S>(
+    public_key: &mut RlwePublicKeyRef<S>,
+    secret_key: &GlweSecretKeyRef<S>,
+    glwe: &GlweDef,
+) where
+    S: TorusOps,
+{
+    assert_eq!(glwe.dim.size.0, 1);
+
+    allocate_scratch_ref!(pt, PolynomialRef<Torus<S>>, (glwe.dim.polynomial_degree));
+    pt.clear();
+
+    encrypt_glwe_ciphertext_secret(public_key.zero_encryption_mut(), pt, secret_key, glwe);
+}
+
+/// Encode and encrypt `msg` into a GLWE ciphertext using the given RLWE public key.
+///
+/// # Remarks
+/// RLWE is a special case of GLWE where N is a power of 2 as normal, but size is 1.
+///
+/// # Panics
+/// If `glwe.dim.size != 1` or if the parameters don't match `ct`, `msg`, or `public_key`.
+pub fn rlwe_encode_encrypt_public<S>(
+    ct: &mut GlweCiphertext<S>,
+    msg: &PolynomialRef<S>,
+    public_key: &RlwePublicKeyRef<S>,
+    plaintext_bits: &PlaintextBits,
+    glwe: &GlweDef,
+) -> RlwePublicEncryptionRandomness<S>
+where
+    S: TorusOps,
+{
+    allocate_scratch_ref!(encoded, PolynomialRef<Torus<S>>, (msg.degree()));
+
+    for (enc, msg) in encoded.coeffs_mut().iter_mut().zip(msg.coeffs()) {
+        *enc = Torus::encode(*msg, *plaintext_bits);
+    }
+
+    rlwe_encrypt_public(ct, encoded, public_key, glwe)
+}
+
+/// Encrypts an encoded polynomial message using the given [`RlwePublicKey`](RlwePublicKeyRef).
+///
+/// # Algorithm
+/// This method uses the double-lwe trick, which is efficient but introduces
+/// more noise than Regev's trick (i.e. summing many encryptions of zero).
```

Comment:
If you have a reference for this in the literature it would be helpful to add. Or add the equation for how decrypting `(p0 * u + e0, m + p1 * u + e1)` works.